src/test/ui/drop/drop_order.rs

```diff
@@ -116,6 +118,58 @@ impl DropOrderCollector {
         }
     }
 
+    fn let_chain(&self) {
+        // take the "then" branch
+        if self.option_loud_drop(2).is_some() // 2
+            && self.option_loud_drop(1).is_some() // 1
+            && let Some(_d) = self.option_loud_drop(4) { // 4
+            self.print(3); // 3
+        }
+
+        // take the "else" branch
+        if self.option_loud_drop(6).is_some() // 2
+            && self.option_loud_drop(5).is_some() // 1
+            && let None = self.option_loud_drop(7) { // 3
+            unreachable!();
+        } else {
+            self.print(8); // 4
+        }
+
+        // let exprs interspersed
+        if self.option_loud_drop(9).is_some() // 1
+            && let Some(_d) = self.option_loud_drop(13) // 5
+            && self.option_loud_drop(10).is_some() // 2
+            && let Some(_e) = self.option_loud_drop(12) { // 4
+            self.print(11); // 3
+        }
+
+        // let exprs first
+        if let Some(_d) = self.option_loud_drop(18) // 5
+            && let Some(_e) = self.option_loud_drop(17) // 4
+            && self.option_loud_drop(14).is_some() // 1
+            && self.option_loud_drop(15).is_some() { // 2
```

Comment:
The drop order is a bit surprising. IIUC, we start by dropping the first regular conditions in *reverse* source order, then other regular conditions in *forward* source order, then the block, then the let conditions. 

Why the forward/reverse mix?

Reply:
Yeah that's definitely a quirk. It's pretty much because for that condition we go from an AST like

```rust
// Fake notation - the numbers correspond to the argument to `option_loud_drop`
Binary(And, Binary(And, Binary(And, Let(18), Let(17)), MethodCall(14)), MethodCall(15)
```

to

```rust
Binary(And, Binary(And, Binary(And, Let(18), Let(17)), DropTemps(MethodCall(14))), DropTemps(MethodCall(15)))
```

If we could group
```rust
self.option_loud_drop(14).is_some() // 1
            && self.option_loud_drop(15).is_some() // 2
```
into the same `DropTemps` then they would drop in reverse source order, but due to the nesting, I couldn't see a way to wrap them in the same `DropTemps` without the let conditions being lumped in.

If the regular conditions come first, like

```rust
self.option_loud_drop(14).is_some()
    && self.option_loud_drop(15).is_some()
    && let Some(_d) = self.option_loud_drop(18)
    && let Some(_e) = self.option_loud_drop(17)
```

we can (and do) wrap them in the same `DropTemps`, and so they drop in reverse source order.

Reply:
FTR there is also pretty weird behaviour for vanilla non-let if chains. Drop behaviour is currently like:

```
        if self.option_loud_drop(5).is_some()
            && self.option_loud_drop(1).is_some()
            && self.option_loud_drop(2).is_some()
            && self.option_loud_drop(3).is_some()
            && self.option_loud_drop(4).is_some() {
            self.print(6);
        }
```

Same goes if you put the expression into a `let _= ...`, or replace the `&&` with a `||` (while changing the is_some to is_none so that all partial exprs get evaluated). So second sub-expression is first, then it goes in normal direction, until the end. Then comes the first sub-expression. Non-short-circuiting & and | drop more consistently in inverse order (5,4,3,2,1). I'm not sure how to resolve this, and how to integrate this with let chains. There is the danger of furthering technical debt on one hand, on the other hand there is the danger of adding exceptions.

_____

A bit unrelated: *Generally* I think that dropping the temporaries of the if let expression after the `else` block is entered was a mistake. Ideally it should be changed to always pre-drop, like what if is doing. I think let chains give us a good opportunity to switch to the new behaviour, at least for let chains, and I'm very happy that currently this is what this PR seems to be doing: drop all temporaries, from let or from an expression, before the `else` block is entered. Just as a positive feedback :).